practice/rain.py

```diff
@@ -4,139 +4,99 @@
-#     :return:
-#     """
-#     days_without_value = find_days_without_value(lookup_table)
-#     filtered_table = {lookup_table.pop(item) for item in list(lookup_table) if item in list(days_without_value.keys())}
-#     return filtered_table
-
-
-def find_days_without_value(lookup_table):
+    year_to_amts= {}
+    for date, amt in date_to_amt.items():
+        year = date[-4:]
+        if year not in year_to_amts:
+            year_to_amts[year] = []
+        year_to_amts[year] += [amt]
+    return year_to_amts
+
+def find_year_with_most_rain(year_to_amts):
+    """Compares all items and returns the year with the highest value.
+
+    >>> find_year_with_most_rain({'2011': [0, 1, 2, 3], '2012': [2, 3, 4, 5]})
+    '2012'
     """
-    >>> find_days_without_value({'15-DEC-2010': '-', '24-JUL-2009': '0'})
-    {'B': '-'}
+    total_rain_by_year = sum_yearly_amounts(year_to_amts)
+    return max(total_rain_by_year.items(), key=operator.itemgetter(1))[0]
 
-    :param lookup_table:
-    :return:
-    """
-    days_without_value = {k: v for k, v in lookup_table.items() if '-' in v}
-    return days_without_value
-
-def correct_days_without_value(days_without_value):
-    for key in days_without_value:
-        days_without_value[key] = '0'
-    return days_without_value
-
-def sort_table_by_year(filtered_table): # Find a sexier solution.
-    table_sorted_by_year = [[][][][][][][][][][][][][][][]]
-    for key in filtered_table:
-        if key[10:11] == '02':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '03':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '04':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '05':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '06':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '07':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '08':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '09':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '10':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '11':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '12':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '13':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '14':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '15':
-            table_sorted_by_year[0] += [key]
-        if key[10:11] == '16':
-            table_sorted_by_year[0] += [key]
-# 2. Define
+def sum_yearly_amounts(year_to_amts):
+    """Maps the dictionary and sums the values of all items.
 
+    >>> sorted(sum_yearly_amounts({'2011': [0, 1, 2, 3], '2012': [2, 3, 4, 5]}).items())
+    [('2011', 6), ('2012', 14)]
+    """
+    year_to_amts.update({key: sum(value) for key, value in year_to_amts.items()})
+    return year_to_amts
```

Comment:
Why are you updating the input dict? Just return the dict comprehension?

1. That modifies the input in the calling function, which is a no-no and confusing.
1. `dict.update()` unions keys so it only produces a new dict by luck of sorts.